File: physics-core/src/collider/spatial_hash.rs
// physics-core/src/collider/spatial_hash.rs
use crate::data::PhysicsData;

pub struct SpatialHash {
    cell_size: f32,
    table_size: usize,
    table: Vec<i32>, // Stores the 'head' particle index for each cell
    next: Vec<i32>,  // Stores the 'next' particle index in the chain
    query_ids: Vec<usize>, // Temp buffer for neighbors
}

impl SpatialHash {
    pub fn new(count: usize, cell_size: f32) -> Self {
        // Table size should be ~2x particle count and Prime to reduce collisions
        let table_size = next_prime(count * 2);

        SpatialHash {
            cell_size,
            table_size,
            table: vec![-1; table_size],
            next: vec![-1; count],
            query_ids: Vec::with_capacity(64),
        }
    }

    pub fn solve(&mut self, data: &mut PhysicsData) {
        let thickness = 0.005; // 0.5cm separation
        let stiffness = 0.1;   // Soft repulsion

        // 1. Clear Table
        self.table.fill(-1);
        self.next.fill(-1);

        // 2. Build Hash Grid
        for i in 0..data.count {
            let idx = i * 3;
            let h = self.hash(
                data.positions[idx],
                data.positions[idx+1],
                data.positions[idx+2]
            );

            // Insert into linked list:
            // Point current particle to what was previously at head
            self.next[i] = self.table[h];
            // Set current particle as new head
            self.table[h] = i as i32;
        }

        // 3. Solve Collisions
        for i in 0..data.count {
            if data.inv_mass[i] == 0.0 { continue; }

            let idx_a = i * 3;
            let px = data.positions[idx_a];
            let py = data.positions[idx_a+1];
            let pz = data.positions[idx_a+2];

            // Find neighbors
            self.query(px, py, pz, thickness);

            for &j in &self.query_ids {
                if i == j { continue; }

                let idx_b = j * 3;
                let dx = px - data.positions[idx_b];
                let dy = py - data.positions[idx_b+1];
                let dz = pz - data.positions[idx_b+2];

                let dist_sq = dx*dx + dy*dy + dz*dz;

                if dist_sq < thickness * thickness && dist_sq > 1e-6 {
                    let dist = dist_sq.sqrt();
                    let correction = (thickness - dist) * 0.5 * stiffness;

                    let nx = dx / dist;
                    let ny = dy / dist;
                    let nz = dz / dist;

                    let w1 = data.inv_mass[i];
                    let w2 = data.inv_mass[j];
                    let w_sum = w1 + w2;

                    if w_sum > 0.0 {
                        let s = correction / w_sum;

                        if w1 > 0.0 {
                            data.positions[idx_a] += nx * s * w1;
                            data.positions[idx_a+1] += ny * s * w1;
                            data.positions[idx_a+2] += nz * s * w1;
                        }
                        if w2 > 0.0 {
                            data.positions[idx_b] -= nx * s * w2;
                            data.positions[idx_b+1] -= ny * s * w2;
                            data.positions[idx_b+2] -= nz * s * w2;
                        }
                    }
                }
            }
        }
    }

    fn hash(&self, x: f32, y: f32, z: f32) -> usize {
        let xi = (x / self.cell_size).floor() as i32;
        let yi = (y / self.cell_size).floor() as i32;
        let zi = (z / self.cell_size).floor() as i32;

        // Large primes for hashing
        let h = (xi.wrapping_mul(73856093)) ^
                (yi.wrapping_mul(19349663)) ^
                (zi.wrapping_mul(83492791));

        (h.abs() as usize) % self.table_size
    }

    fn query(&mut self, x: f32, y: f32, z: f32, r: f32) {
        self.query_ids.clear();

        let x0 = ((x - r) / self.cell_size).floor() as i32;
        let x1 = ((x + r) / self.cell_size).floor() as i32;
        let y0 = ((y - r) / self.cell_size).floor() as i32;
        let y1 = ((y + r) / self.cell_size).floor() as i32;
        let z0 = ((z - r) / self.cell_size).floor() as i32;
        let z1 = ((z + r) / self.cell_size).floor() as i32;

        for xi in x0..=x1 {
            for yi in y0..=y1 {
                for zi in z0..=z1 {
                    let h = (xi.wrapping_mul(73856093) ^
                             yi.wrapping_mul(19349663) ^
                             zi.wrapping_mul(83492791)).abs() as usize % self.table_size;

                    let mut id = self.table[h];
                    while id != -1 {
                        self.query_ids.push(id as usize);
                        id = self.next[id as usize];
                    }
                }
            }
        }
    }
}

fn next_prime(mut n: usize) -> usize {
    loop {
        let mut is_prime = true;
        let limit = (n as f64).sqrt() as usize;
        for i in 2..=limit {
            if n % i == 0 {
                is_prime = false;
                break;
            }
        }
        if is_prime { return n; }
        n += 1;
    }
}

File: verlet-rapier-xpbd-simulation/src/archives/v3_xpbd/engine/core/SpatialHash.ts
// src/v3/engine/collision/SpatialHash.ts
import { PhysicsData } from '../core/PhysicsData';

export class SpatialHash {
    private cellSize: number;
    private tableSize: number;
    private table: Int32Array;
    private next: Int32Array;
    private queryIds: number[] = [];

    constructor(count: number, cellSize: number = 0.015) {
        this.cellSize = cellSize;
        this.tableSize = this.getNextPrime(count * 2);
        this.table = new Int32Array(this.tableSize);
        this.next = new Int32Array(count);
    }

    private getNextPrime(n: number): number {
        while (true) {
            let isPrime = true;
            for (let i = 2; i * i <= n; i++) {
                if (n % i === 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) return n;
            n++;
        }
    }

    public solve(data: PhysicsData) {
        const count = data.count;
        const pos = data.positions;
        const invMass = data.invMass;
        const thickness = 0.01; // 1cm thickness
        const stiffness = 0.2;

        // 1. Update Hash Table
        this.table.fill(-1);
        this.next.fill(-1);

        for (let i = 0; i < count; i++) {
            const x = pos[i * 3];
            const y = pos[i * 3 + 1];
            const z = pos[i * 3 + 2];
            const h = this.hash(x, y, z);
            this.next[i] = this.table[h];
            this.table[h] = i;
        }

        // 2. Solve Collisions
        for (let i = 0; i < count; i++) {
            if (invMass[i] === 0) continue;

            const idxA = i * 3;
            const px = pos[idxA];
            const py = pos[idxA + 1];
            const pz = pos[idxA + 2];

            this.query(px, py, pz, thickness);

            for (let k = 0; k < this.queryIds.length; k++) {
                const j = this.queryIds[k];
                if (i === j) continue;

                const idxB = j * 3;
                const dx = px - pos[idxB];
                const dy = py - pos[idxB + 1];
                const dz = pz - pos[idxB + 2];
                const distSq = dx * dx + dy * dy + dz * dz;

                if (distSq < thickness * thickness && distSq > 0.000001) {
                    const dist = Math.sqrt(distSq);
                    const correction = (thickness - dist) * 0.5 * stiffness;

                    const nx = dx / dist;
                    const ny = dy / dist;
                    const nz = dz / dist;

                    const wA = invMass[i];
                    const wB = invMass[j];
                    const wSum = wA + wB;

                    if (wSum > 0) {
                        const s = correction / wSum;
                        if (wA > 0) {
                            pos[idxA] += nx * s * wA;
                            pos[idxA + 1] += ny * s * wA;
                            pos[idxA + 2] += nz * s * wA;
                        }
                        if (wB > 0) {
                            pos[idxB] -= nx * s * wB;
                            pos[idxB + 1] -= ny * s * wB;
                            pos[idxB + 2] -= nz * s * wB;
                        }
                    }
                }
            }
        }
    }

    private query(x: number, y: number, z: number, radius: number) {
        this.queryIds.length = 0;
        const x0 = Math.floor((x - radius) / this.cellSize);
        const y0 = Math.floor((y - radius) / this.cellSize);
        const z0 = Math.floor((z - radius) / this.cellSize);
        const x1 = Math.floor((x + radius) / this.cellSize);
        const y1 = Math.floor((y + radius) / this.cellSize);
        const z1 = Math.floor((z + radius) / this.cellSize);

        for (let xi = x0; xi <= x1; xi++) {
            for (let yi = y0; yi <= y1; yi++) {
                for (let zi = z0; zi <= z1; zi++) {
                    const h = this.hashCoords(xi, yi, zi);
                    let id = this.table[h];
                    while (id !== -1) {
                        this.queryIds.push(id);
                        id = this.next[id];
                    }
                }
            }
        }
    }

    private hash(x: number, y: number, z: number): number {
        const xi = Math.floor(x / this.cellSize);
        const yi = Math.floor(y / this.cellSize);
        const zi = Math.floor(z / this.cellSize);
        return this.hashCoords(xi, yi, zi);
    }

    private hashCoords(xi: number, yi: number, zi: number): number {
        const h = (xi * 73856093) ^ (yi * 19349663) ^ (zi * 83492791);
        return (h & 0x7fffffff) % this.tableSize;
    }
}